File: samples/python/retrieval-agent/src/services/message_route.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from __future__ import annotations

from typing import Awaitable, Callable

from .card import create_source_card
from .retrieval_client import RetrievalResult, RetrievalStatus

BUILD_GENIE_RESPONSES = {
    RetrievalStatus.NOT_SIGNED_IN: "Please sign in to Microsoft 365, then ask your Build question again.",
    RetrievalStatus.NO_RESULTS: "I couldn't find Build session information in the configured SharePoint site. Check the site URL, document permissions, and indexing, then try a more specific question.",
    RetrievalStatus.SERVICE_UNAVAILABLE: "I couldn't retrieve Build session information right now. Please try again later.",
}


def grounded_answer(result: RetrievalResult) -> str:
    sources = "\n\n".join(
        f"{item.title}\n{item.extract}\nSource: {item.web_url}" for item in result.items
    )
    return f"Here is what I found in the configured SharePoint site:\n\n{sources}"


async def handle_build_genie_message(
    result: RetrievalResult,
    send_text: Callable[[str], Awaitable[object]],
    send_source_card: Callable[[dict], Awaitable[object]],
) -> None:
    if result.status is not RetrievalStatus.SUCCESS:
        await send_text(BUILD_GENIE_RESPONSES[result.status])
        return
    await send_text(grounded_answer(result))
    await send_source_card(create_source_card(result.items))

File: samples/nodejs/retrieval-agent/src/services/messageRoute.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { createSourceCard } from './card.js'
import type { RetrievalResult, RetrievalStatus } from './retrievalClient.js'

export const buildGenieResponses: Record<Exclude<RetrievalStatus, 'success'>, string> = {
  notSignedIn: 'Please sign in to Microsoft 365, then ask your Build question again.',
  noResults: "I couldn't find Build session information in the configured SharePoint site. Check the site URL, document permissions, and indexing, then try a more specific question.",
  serviceUnavailable: "I couldn't retrieve Build session information right now. Please try again later."
}

export function groundedAnswer (result: RetrievalResult): string {
  return `Here is what I found in the configured SharePoint site:\n\n${result.items.map(item => `${item.title}\n${item.extract}\nSource: ${item.webUrl}`).join('\n\n')}`
}

export async function handleBuildGenieMessage (
  result: RetrievalResult,
  sendText: (text: string) => Promise<unknown>,
  sendSourceCard: (card: object) => Promise<unknown>
): Promise<void> {
  if (result.status !== 'success') {
    await sendText(buildGenieResponses[result.status])
    return
  }

  await sendText(groundedAnswer(result))
  await sendSourceCard(createSourceCard(result.items))
}
